e2e/internal/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,341 @@
+						return false
+					}
+				}
+			}
+		}
+
+		if controlNodeCount != wantControlNodeCount {
+			fmt.Printf("Want %d control nodes but got %d\n", wantControlNodeCount, controlNodeCount)
+			return false
+		}
+		if workerNodeCount != wantWorkerNodeCount {
+			fmt.Printf("Want %d worker nodes but got %d\n", wantWorkerNodeCount, workerNodeCount)
+			return false
+		}
+
+		return true
+	}, time.Minute*30, time.Minute)
+}
+
+type imageMeasurements struct {
+	Cmdline       string `json:"cmdline"`
+	CmdlineSHA256 string `json:"cmdline-sha256"`
+	EFIstages     []struct {
+		Name   string `json:"name"`
+		SHA256 string `json:"sha256"`
+	} `json:"efistages"`
+	InitrdSHA256 string                       `json:"initrd-sha256"`
+	Measurements map[string]map[string]string `json:"measurements"`
+}
+
+type upgradeInfo struct {
+	measurements map[uint32]string
+	shortPath    string
+	wantImage    string
+}
+
+func fetchLatestNightlyUpgradeInfo(csp cloudprovider.Provider) (upgradeInfo, error) {
+	info := upgradeInfo{
+		measurements: make(map[uint32]string),
+	}
+	versionsClient := fetcher.NewFetcher()
+
+	latest, err := versionsClient.FetchVersionLatest(context.Background(), versionsapi.Latest{
+		Ref:    "main",
+		Stream: "nightly",
+		Kind:   versionsapi.VersionKindImage,
+	})
+	if err != nil {
+		return upgradeInfo{}, err
+	}
+	info.shortPath = latest.ShortPath()
+
+	ver := versionsapi.Version{
+		Ref:     latest.Ref,
+		Stream:  latest.Stream,
+		Version: latest.Version,
+		Kind:    latest.Kind,
+	}
+	artifactURL := ver.ArtifactURL()
+
+	measurementsURL, err := url.JoinPath(artifactURL, "image/csp", strings.ToLower(csp.String()), "measurements.image.json")
+	if err != nil {
+		return upgradeInfo{}, err
+	}
+
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, measurementsURL, http.NoBody)
+	if err != nil {
+		return upgradeInfo{}, err
+	}
+	fmt.Printf("Fetching measurements from: %v\n", req.URL)
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return upgradeInfo{}, err
+	}
+	defer resp.Body.Close()
+
+	var imageMeasurements imageMeasurements
+	err = json.NewDecoder(resp.Body).Decode(&imageMeasurements)
+	if err != nil {
+		return upgradeInfo{}, err
+	}
+
+	for key, value := range imageMeasurements.Measurements {
+		idx, err := strconv.Atoi(key)
+		if err != nil {
+			return upgradeInfo{}, err
+		}
+		info.measurements[uint32(idx)] = value["expected"]
+	}
+
+	imageInfo, err := versionsClient.FetchImageInfo(context.Background(), versionsapi.ImageInfo{
+		Ref:     latest.Ref,
+		Stream:  latest.Stream,
+		Version: latest.Version,
+	})
+	if err != nil {
+		return upgradeInfo{}, err
+	}
+
+	switch csp {
+	case cloudprovider.GCP:
+		info.wantImage = imageInfo.GCP["sev-es"]
+	case cloudprovider.Azure:
+		info.wantImage = imageInfo.Azure["cvm"]
+	case cloudprovider.AWS:
+		info.wantImage = imageInfo.AWS["eu-central-1"]
+	default:
+		return upgradeInfo{}, errors.New("finding wanted image")
+	}
+
+	return info, nil
```

Comment:
I think this could be replaced with a call to [getCompatibleImageMeasurements](https://github.com/edgelesssys/constellation/blob/main/cli/internal/cmd/upgradeplan.go#L200)? The `images` slice is expected to be a slice of short paths.

Reply:
Good suggestion to reuse existing code! 

I did look into it, but it seems the suggested version uses [measurements.json](https://cdn.confidential.cloud/constellation/v1/ref/-/stream/stable/v2.5.0/image/csp/gcp/measurements.json) which is only available in a released version. I keep the code based on [measurements.image.json](https://cdn.confidential.cloud/constellation/v1/ref/-/stream/stable/v2.5.0/image/csp/gcp/measurements.image.json) since it is available for all branches. 

I will have another look why we have two files in a seperate PR.